test/jdk/java/foreign/TestMismatch.java

```diff
@@ -0,0 +1,236 @@
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.IntFunction;
+import jdk.incubator.foreign.MemoryAddress;
+import jdk.incubator.foreign.MemoryLayouts;
+import jdk.incubator.foreign.MemorySegment;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import static java.lang.System.out;
+import static jdk.incubator.foreign.MemorySegment.READ;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertThrows;
+
+public class TestMismatch {
+
+    final static VarHandle BYTE_HANDLE = MemoryLayouts.JAVA_BYTE.varHandle(byte.class);
+
+    // stores a increasing sequence of values into the memory of the given segment
+    static MemorySegment initializeSegment(MemorySegment segment) {
+        MemoryAddress addr = segment.baseAddress();
+        for (int i = 0 ; i < segment.byteSize() ; i++) {
+            BYTE_HANDLE.set(addr.addOffset(i), (byte)i);
+        }
+        return segment;
+    }
+
+    @Test(dataProvider = "slices")
+    public void testSameValues(MemorySegment ss1, MemorySegment ss2) {
+        out.format("testSameValues s1:%s, s2:%s\n", ss1, ss2);
+        MemorySegment s1 = initializeSegment(ss1);
+        MemorySegment s2 = initializeSegment(ss2);
+
+        if (s1.byteSize() == s2.byteSize()) {
+            assertEquals(s1.mismatch(s2), -1);  // identical
+            assertEquals(s2.mismatch(s1), -1);
+        } else if (s1.byteSize() > s2.byteSize()) {
+            assertEquals(s1.mismatch(s2), s2.byteSize());  // proper prefix
+            assertEquals(s2.mismatch(s1), s2.byteSize());
+        } else {
+            assert s1.byteSize() < s2.byteSize();
+            assertEquals(s1.mismatch(s2), s1.byteSize());  // proper prefix
+            assertEquals(s2.mismatch(s1), s1.byteSize());
+        }
+    }
+
+    @Test(dataProvider = "slices")
+    public void testDifferentValues(MemorySegment s1, MemorySegment s2) {
+        out.format("testDifferentValues s1:%s, s2:%s\n", s1, s2);
+        s1 = initializeSegment(s1);
+        s2 = initializeSegment(s2);
+
+        for (long i = s2.byteSize() -1 ; i >= 0; i--) {
+            long expectedMismatchOffset = i;
+            BYTE_HANDLE.set(s2.baseAddress().addOffset(i), (byte) 0xFF);
+
+            if (s1.byteSize() == s2.byteSize()) {
+                assertEquals(s1.mismatch(s2), expectedMismatchOffset);
+                assertEquals(s2.mismatch(s1), expectedMismatchOffset);
+            } else if (s1.byteSize() > s2.byteSize()) {
+                assertEquals(s1.mismatch(s2), expectedMismatchOffset);
+                assertEquals(s2.mismatch(s1), expectedMismatchOffset);
+            } else {
+                assert s1.byteSize() < s2.byteSize();
+                var off = Math.min(s1.byteSize(), expectedMismatchOffset);
+                assertEquals(s1.mismatch(s2), off);  // proper prefix
+                assertEquals(s2.mismatch(s1), off);
+            }
+        }
+    }
```

Comment:
How important is it that these tests operate on slices? Looking at the test code, it could have worked equally well if the input parameters were just two sizes, and then you did an explicit allocation (or maybe also receive a segment factory from the provider, so that you can test different segment kinds).

Reply:
Originally I had a version of the test that did compare specific segments, but it didn't scale well to different sizes and kinds ( we need to test both above and below the 8 byte threshold ). I removed a number of slice sizes, which greatly reduces the combinations. This may be enough, or I can certainly revisit the test's structure.